Adds basic NetMessage based examine system (review and comments)

drneuris opened the pull request:
# Pull Request Template

### Purpose
Initially addresses #2340 and #2643 - will close in a later PR when I sort out tiles and construction. Hang in there!

An initial attempt to tackle shift examine. **Disclaimer**: _I'm definitely not a unity programmer and this is my first contribution, so if it's shite, burn it to the ground. However it seemed like a cool way to learn and get something useful out of it. Hopefully it can be used._

Added Shift Click detection in appropriate controllers
Added IExaminable interface (see below) - One method examine(), returns whatever is appropriate (server process() needs to take care of it)
Shift Clicking sends a (new) RequestExamineMessage to server, who can process() as it sees fit. At the moment it looks at the target object for anything extending IExaminable and calls examine() on them, uses the returned string to assemble a final string, then uses Chat.AddExamineMsgFromServer to pass back the string. This keeps the logic server side and somewhat sane.
Any clickable (for now not tile/matrix based) can have any of its components extend IExaminable to automatically append text to returned message.

See below for implemented features/changes. 

**Minimum functionality for final review and merge request:** 

- [x] attributes component should be examinable and return minimal description
- [x] add consciousness/dead/is breathing(?) data in examine() for livinghelthbehaviour
- [x] Add report to Integrity component for damaged items
- [x] Adapt Equipment component to provide and hide identity as needed (when no id card present).
- [x] use article name in descriptions
- [x] Refactored information flow from ID card to various info outlets (OnHover, shiftclicking a player, shiftclicking ID card item should now refer to same source of information and criteria for ID on belt)
- [x] This means that no id -> no name, also for mouse hover-text

**Desirable functionality for next iteration (After careful consideration, this is going to a separate PR/issue) loosely ordered by increasing complexity**
- adding functionality for in-progress construction to provide feedback as per #2340 (reinf wall and computer construction) - gotta look into construction code, PR is supposed to solve that specifically but can't really give eta on it cause I don't know construction yet. Comments?
- add more medical flavor (*he has severe/moderate bruising! he is bleeding! he has minor burns.* etc)
- Let examine get clothes on people at very least (in itemmanager probably have an array of named itemslots in to collect and send)
- Let server calculate distance between interactor and interactee to allow distance based information details (don't want to see small cuts on people if 10 tiles away)
- inclusion of right click items logic to streamline examine behavior
- refactor existing OnExamine behaviours to be compatible/get rid of it entirely
- At least an initial implementation of security checks / visibility checks (also on the client side to avoid sending unnecessary messages)
- tile based examine implementation (cause i don't know enough about tiles right now)
- decide/figure out mechanics for showing name based on face visibility (Right now no id on belt -> unknown, but it doesn't make complete sense).

There is a pretty big limitation right now in how Equipment doesn't have access to ID slot because players are not observers of the ID slot by default. I think it should be changed to have everyone be an observer of everyone else's ID slot, OR (maybe better?) with a playerappearancemessage somehow? I'm not sure, this will probably have to be solved before merge or added as an issue for later. WORKAROUND: have hover text always show name like it was before.

Logistical/documentation tasks:
- Create separate issues to track what components need to get IExaminable support in the future and what will be done in the future (do after merge?)
- make sure necessary files are pushed - .meta etc - can anyone pitch in in comments?
- comments/todos
- create wiki page with basic details to allow anyone to extend


### Notes:
All very much wip, any and all feedback is welcome. Style could be better, and i'm probably forgetting many guidelines. Hope it's not too bad!

Example:
![image](https://user-images.githubusercontent.com/3104656/74886340-2202a100-5378-11ea-9212-564248f98db8.png)



Targeting 20 feb for completion/review.

### Please make sure you have followed the self checks below before submitting a PR:

- [x] - Code is sufficiently commented
- [x] - Code is indented with tabs and not spaces
- [x] - The PR does not include any unnecessary .meta, .prefab or <b>.unity (scene) changes</b>
- [x] - The PR does not bring up any new compile errors
- [x] - The PR has been tested in editor
- [x] - Any new files are named using PascalCase (to avoid issues on case sensitive file systems)
- [x] - Any new / changed components follow the [Component Development Checklist](https://github.com/unitystation/unitystation/wiki/Component-Development-Checklist)
- [x] - Any new objects / items follow the [Creating Items and Objects Guide](https://github.com/unitystation/unitystation/wiki/Creating-Items-and-Objects%3A-Now-With-Prefab-Variants) (especially concerning the use of prefab variants)
- [x] - The PR has been tested in multiplayer (with 2 clients and late joiners, if applicable)
- [x] - The PR has been tested with round restarts.

drneuris committed:
Examine v1

goman0 commented on UnityProject/Assets/Scripts/Health/Living/LivingHealthBehaviour.cs:
```diff
@@ -752,6 +752,53 @@ private void OnDrawGizmos()
+
+		var cs = this.GetComponentInParent<PlayerScript>()?.characterSettings;
+		if (cs != null)
+		{
```
this block should be indented

goman0 commented on UnityProject/Assets/Scripts/Input System/MouseInputController.cs:
```diff
@@ -128,8 +128,17 @@ private void CheckMouseInput()
+
+            //check the shift, alt click and throw, which doesn't have any special logic
+            
+            if (CheckAltClick()) return;
```
extra space here

goman0 commented on UnityProject/Assets/Scripts/Input System/MouseInputController.cs:
```diff
@@ -128,8 +128,17 @@ private void CheckMouseInput()
+			{
+				//like above, send shift-click request, then do nothing else.  
+				CheckShiftClick(); 
+				
```
extra line here

goman0 commented on UnityProject/Assets/Scripts/Input System/MouseInputController.cs:
```diff
@@ -473,8 +482,24 @@ private MouseDraggable GetDraggable()
-	private bool CheckAltClick()
+    /// <summary>
+    /// Fires if shift is pressed on click, initiates examine. Assumes inanimate object, but upgrades to checking health if living, and id if target has 
+	/// storage and an ID card in-slot.
```
this is mis-indented

goman0 commented on UnityProject/Assets/Scripts/Input System/MouseInputController.cs:
```diff
@@ -473,8 +482,24 @@ private MouseDraggable GetDraggable()
+	/// storage and an ID card in-slot.
+    /// </summary>
+    /// <returns>false if no shiftclick, true if something was examined (for future chatbox related events, or gui)</returns>
+    private bool CheckShiftClick()
```
this block mixes tabs & spaces

goman0 commented on UnityProject/Assets/Scripts/Input System/MouseInputController.cs:
```diff
@@ -473,8 +482,24 @@ private MouseDraggable GetDraggable()
+        return false;
+    }
+
+    private bool CheckAltClick()
```
weird indentation here

goman0 commented on UnityProject/Assets/Scripts/Inventory/ItemStorage.cs:
```diff
@@ -16,8 +16,7 @@
 /// </summary>
-public class ItemStorage : MonoBehaviour, IServerLifecycle, IServerInventoryMove, IClientInventoryMove,
-	IClientDespawn
+public class ItemStorage : MonoBehaviour, IServerLifecycle, IServerInventoryMove, IClientInventoryMove,	IClientDespawn
```
this could probably be split up over multiple lines at this point, otherwise I would remove the extra tab between `IClientInventoryMove` and `IClientDespawn`

goman0 commented on UnityProject/Assets/Scripts/Messages/Client/Interaction/RequestExamineMessage.cs:
```diff
@@ -0,0 +1,162 @@
+
```
extra \n here

goman0 commented on UnityProject/Assets/Scripts/Health/Living/LivingHealthBehaviour.cs:
```diff
@@ -752,6 +752,53 @@ private void OnDrawGizmos()
+		// Assume animal
+		string pronoun = "It";
+
+		var cs = this.GetComponentInParent<PlayerScript>()?.characterSettings;
```
explicit `this` is not necessary

goman0 commented on UnityProject/Assets/Scripts/Messages/Client/Interaction/RequestExamineMessage.cs:
```diff
@@ -0,0 +1,162 @@
+		var examinables = NetworkObject.GetComponents<IExaminable>();
+		foreach (var examinable in examinables)
+		{
+		msg += $"{examinable.examine()}";
```
this should be indented

goman0 commented on UnityProject/Assets/Scripts/Messages/Client/Interaction/RequestExamineMessage.cs:
```diff
@@ -0,0 +1,162 @@
+
+		
+		// Send the message.
+		GameObject  go = SentByPlayer.GameObject;
```
It's probably not necessary to store this in a temporary variable

goman0 commented on UnityProject/Assets/Scripts/Input System/MouseInputController.cs:
```diff
@@ -473,8 +482,24 @@ private MouseDraggable GetDraggable()
+	/// storage and an ID card in-slot.
+    /// </summary>
+    /// <returns>false if no shiftclick, true if something was examined (for future chatbox related events, or gui)</returns>
+    private bool CheckShiftClick()
```
does this function `return true;` anywhere?

drneuris replied:
changed to void

drneuris committed:
Small fixes

Fixes goman's comments; bunch of indentation and whitespaces are from before my edits, but still, thanks for the input!

drneuris committed:
Moar features

drneuris committed:
Updates id card behavior - hover text - adds some helper functions

Still not able to get ID card deets from Equipment when calling from client. This should be easily fixable but can't figure it out yet.

drneuris committed:
Final touches

Changed descriptions returned from Attributes to match mouseover text.
Fixed extra LF on examine messages.

fomalsd commented on UnityProject/Assets/Scripts/Messages/Client/Interaction/RequestExamineMessage.cs:
```diff
@@ -42,10 +42,16 @@ public override IEnumerator Process()
+
+		if (msg.Length > 0)
+		{
+			msg = msg.Substring(0, msg.Length-2);
```
what's that for?

fomalsd replied:
it'll also give an exception if `msg.Length` is 1

drneuris replied:
It'll never be, but i changed it all the same; it was to remove the newline for message sent by the last examinable, to avoid wasting space in chat, but i wrote it at 3am so it's a bit stupid. It's a for loop now!

drneuris committed:
Update RequestExamineMessage.cs

fomalsd commented on UnityProject/Assets/Scripts/Messages/Client/Interaction/RequestExamineMessage.cs:
```diff
@@ -43,16 +43,19 @@ public override IEnumerator Process()
 		{
-			msg = msg.Substring(0, msg.Length-2);
+			examinable = examinables[i];
+			msg += $"{examinable.Examine()}";
```
You don’t need to do string interpolation and just add examinable.Examine() without quotes, but that’s a minor nitpick. Also I’ve got a feeling that this whole loop could be avoided by simply using string.Join...

fomalsd reviewed (approved):
merging this! shift+click now works!
